pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,341 @@
+	Region          string `yaml:"region"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	SecretAccessKey string `yaml:"secret_access_key"`
+}
+
+func (conf *Config) validate() error {
+	if conf.Bucket == "" ||
+		conf.Region == "" ||
+		conf.AccessKeyID == "" ||
+		conf.SecretAccessKey == "" {
+		return errors.New("insufficient bos configuration information")
+	}
+	return nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	var config *Config
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, errors.Wrap(err, "parsing bos configuration")
+	}
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate bos configuration")
+	}
+
+	client := bos.NewClient(&bce.Config{
+		Region: config.Region,
+		Credentials: &bce.Credentials{
+			AccessKeyID:     config.AccessKeyID,
+			SecretAccessKey: config.SecretAccessKey,
+		},
+	})
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		name:   config.Bucket,
+	}
+	return bkt, nil
+}
+
+// Name returns the bucket name for the provider.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.client.DeleteObject(b.name, name, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if _, err := b.client.PutObject(b.name, name, r, nil, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Iter calls f for each entry in the given directory (not recursive.). The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, dirDelim) + dirDelim
+	}
+
+	for object := range b.listObjects(ctx, dir) {
+		if object.err != nil {
+			return object.err
+		}
+		if object.key == "" {
+			continue
+		}
+		if err := f(object.key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getRange(ctx, b.name, name, 0, -1)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getRange(ctx, b.name, name, off, length)
+}
+
+// Exists checks if the given object exists in the bucket.
+// TODO(bplotka): Consider removing Exists in favor of helper that do Get & IsObjNotFoundErr (less code to maintain).
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	obj, err := b.getRange(ctx, b.name, name, 0, 2)
```

Comment:
Just curious, why did you use `getRange` to check the object? Could we use the `GetObjectMeta` API?